velocity_modelling/cvm/velocity3d.py

```diff
@@ -0,0 +1,786 @@
+            Global surface depth data.
+        velo_mod_1d_data : VelocityModel1D
+            1D velocity model data.
+        nz_tomography_data : TomographyData
+            Tomography data.
+        mesh_vector : MeshVector
+            Lat-lon point with multiple depths.
+        in_any_basin_lat_lon : bool
+            Whether this lat-lon point is in any basin.
+        on_boundary : bool
+            Whether this point is on a model boundary.
+        """
+        # Sort by z_indices to preserve depth order
+        order = np.argsort(z_indices)
+        z_indices = z_indices[order]
+        depths = depths[order]
+        submodel_names = submodel_names[order]
+
+        for name in np.unique(submodel_names):
+            mask = submodel_names == name
+            depth_subset = depths[mask]
+            index_subset = z_indices[mask]
+
+            if name == "nan_submod":
+                from velocity_modelling.cvm.submodel import (
+                    nan_submod,
+                )
+
+                nan_submod.main_vectorized(index_subset, self)
+            elif name == "ep_tomography_submod_v2010":
+                from velocity_modelling.cvm.submodel import ep_tomography_submod_v2010
+
+                # Precompute interpolated values for all surfaces and vtype at this (lat, lon)
+                global_surf_read = nz_tomography_data.surfaces[0]["vp"]
+                adjacent_points = AdjacentPoints.find_global_adjacent_points(
+                    global_surf_read.lats,
+                    global_surf_read.lons,
+                    mesh_vector.lat,
+                    mesh_vector.lon,
+                )
+
+                # Precompute interpolated values for all surfaces and vtype
+                num_surfaces = len(nz_tomography_data.surfaces)
+                interpolated_global_surface_values = {
+                    vtype.name: np.zeros(num_surfaces) for vtype in VelocityTypes
+                }
+                for idx in range(num_surfaces):
+                    for vtype in VelocityTypes:
+                        surface = nz_tomography_data.surfaces[idx][vtype.name]
+                        val = interpolate_global_surface(
+                            surface, mesh_vector.lat, mesh_vector.lon, adjacent_points
+                        )
+                        interpolated_global_surface_values[vtype.name][idx] = val
+
+                ep_tomography_submod_v2010.main_vectorized(
+                    index_subset,
+                    depth_subset,
+                    self,
+                    mesh_vector,
+                    nz_tomography_data,
+                    partial_global_surface_depths,
+                    global_params["GTL"],
+                    in_any_basin_lat_lon,
+                    on_boundary,
+                    interpolated_global_surface_values,
+                )
+            elif name == "canterbury1d_v1":
+                from velocity_modelling.cvm.submodel import canterbury1d_v1
```

Comment:
Does this import need to be here?

Reply:
Ideally, I want something that can dynamically locate the module needed to import, like how I loaded "write" module for emod3d or csv. The tricky part is each module needs different arguments etc., so I just did the way it currently is. 

```
   import importlib

    try:
        module_name = f"velocity_modelling.cvm.write.{output_format.lower()}"
        writer_module = importlib.import_module(module_name)
        write_global_qualities = writer_module.write_global_qualities
        logger.log(f"Using {output_format} writer module", logger.DEBUG)
    except ImportError:
        logger.log(f"Unsupported output format: {output_format}", logger.ERROR)
        raise ValueError(f"Unsupported output format: {output_format}")
```